[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ohimjosh/CSC-44800/blob/main/AS01/Miranda_Josh_AS02.ipynb)


In [6]:
import math
import time

def naive(n):
    start_time = time.time()
    #print("Start:", start_time)
    
    s = 0
    print("N:", str(n))
    
    for i in range(0,n):
        s = s + math.sqrt(19.0)
        
    print("Summation:", s)
    
    end_time = time.time()
    #print("End:", end_time)
    
    elapsed_time = end_time - start_time
    print("Elapsed Time:", elapsed_time, "seconds")
    
    exact_sum = n * math.sqrt(19.0)
    print("Exact:", exact_sum)
       
    absolute_error = abs(s-exact_sum)
    print("Absolute Error:", absolute_error)
        
    relative_error = absolute_error/exact_sum
    print("Relative Error:", relative_error)
    
    print("\n")
    
def comp(n):
    start_time = time.time()
    #print("Start:", start_time)
    
    s = 0
    print("N:", str(n))
    
    e = 0
    
    for i in range(0,n):
        temp = s
        y = math.sqrt(19.0) + e
        s = temp + y
        e = (temp - s) + y
    s += e
    print("Summation:", s)
    
    end_time = time.time()
    #print("End:", end_time)
    
    elapsed_time = end_time - start_time
    print("Elapsed Time:", elapsed_time, "seconds")
    
    exact_sum = n * math.sqrt(19.0)
    print("Exact:", exact_sum)
    
    absolute_error = abs(s-exact_sum)
    print("Absolute Error:", absolute_error)
    
    relative_error = absolute_error/exact_sum
    print("Relative Error:", relative_error) 
    
    print("\n")
    
    
def both():    
    N = [10**3,10**4,10**5,10**6,10**7,10**8]
    for x in N:
        print("Naive Algorithm")
        naive(x)
        print("Compensated Algorithm")
        comp(x)
    

print(both())
    


    

Naive Algorithm
N: 1000
Summation: 4358.898943540652
Elapsed Time: 0.0002846717834472656 seconds
Exact: 4358.898943540674
Absolute Error: 2.1827872842550278e-11
Relative Error: 5.007657467006977e-15


Compensated Algorithm
N: 1000
Summation: 4358.898943540674
Elapsed Time: 0.0009000301361083984 seconds
Exact: 4358.898943540674
Absolute Error: 0.0
Relative Error: 0.0


Naive Algorithm
N: 10000
Summation: 43588.98943540103
Elapsed Time: 0.001285552978515625 seconds
Exact: 43588.98943540674
Absolute Error: 5.704350769519806e-09
Relative Error: 1.30866781804449e-13


Compensated Algorithm
N: 10000
Summation: 43588.98943540674
Elapsed Time: 0.0024995803833007812 seconds
Exact: 43588.98943540674
Absolute Error: 0.0
Relative Error: 0.0


Naive Algorithm
N: 100000
Summation: 435889.8943536333
Elapsed Time: 0.008142232894897461 seconds
Exact: 435889.8943540674
Absolute Error: 4.341127350926399e-07
Relative Error: 9.959229170383476e-13


Compensated Algorithm
N: 100000
Summation: 435889.89435406

**Code Discussion**

A class that I have previously taken is CSC-30100 - Scientific Programming. The assignment for this class is to calculate the value $S_{n}$ using two methods and compare them, naive summation and compensated summation. 


![Sn](Sn.png)

In order to test these methods the summations with be perform on N = $10^{3}$, $10^{4}$, $10^{5}$, $10^{6}$, $10^{7}$, and $10^{8}$.



**Naive Summation**

```Python
s = 0
print("N:", str(n))
    
for i in range(0,n):
    s = s + math.sqrt(19.0)
        
print("Summation:", s)
```

Here is the naive algorithm where it will keep adding $\sqrt{19}$ for range 0 to n.



**Compensated Summation**

```Python
s = 0
print("N:", str(n))
e = 0
    
for i in range(0,n):
    temp = s
    y = math.sqrt(19.0) + e
    s = temp + y
    e = (temp - s) + y
s += e
print("Summation:", s)
```
    
The algorithm for compensated summation which is recursive. Similar to the naive algorithm it will add $\sqrt{19}$ for range 0 to n but this time there is a correction term designed to diminish rounding errors, seen in the variable e.



**General Format**

For the two methods they will be compared based of these categories:
* Elapsed Time

```Python
elapsed_time = end_time - start_time
print("Elapsed Time:", elapsed_time, "seconds")
```
  
  
* Exact Value
```Python
exact_sum = n * math.sqrt(19.0)
print("Exact:", exact_sum)
```
  

* Absolute Error
```Python
absolute_error = abs(s-exact_sum)
print("Absolute Error:", absolute_error)
```


* Relative Error
```Python
relative_error = absolute_error/exact_sum
print("Relative Error:", relative_error)
```

**Very Cool Funny Meme**



![Ai Meme](aimeme.jpg)

**Results**

   **Naive Algorithm**

| $N$ | Summation | Elapsed Time | Exact | Absolute Error |  Relative Error      
| :---: | :---: | :---: | :---: | :---: | :---:  
|$10^{3}$| 4358.898943540652  | 0.0 | 4358.898943540674 | 2.1827872842550278e-11 | 5.007657467006977e-15
|$10^{4}$| 43588.98943540103  | 0.0009989738464355469 | 43588.98943540674 | 5.704350769519806e-09 | 1.30866781804449e-13
|$10^{5}$| 435889.8943536333  | 0.018993616104125977 | 435889.8943540674 | 4.341127350926399e-07 | 9.959229170383476e-13
|$10^{6}$| 4358898.943627511  | 0.1379561424255371 | 4358898.943540674 | 8.683744817972183e-05 | 1.9921876901597763e-11
|$10^{7}$| 43588989.44469575  | 1.3758232593536377 | 43588989.43540674 | 0.009289011359214783 | 2.1310453579062438e-10
|$10^{8}$| 435889893.87809396  | 14.282974481582642 | 435889894.3540674 | 0.4759734272956848 | 1.09195793125926e-09


**Compensated Algorithm**

| $N$ | Summation | Elapsed Time | Exact | Absolute Error |  Relative Error      
| :---: | :---: | :---: | :---: | :---: | :---:  
|$10^{3}$| 4358.898943540674  | 0.0 | 4358.898943540674 | 0.0 | 0.0
|$10^{4}$| 43588.98943540674  | 0.003996372222900391 | 43588.98943540674 | 0.0 | 0.0
|$10^{5}$| 435889.8943540674  | 0.022992372512817383 | 435889.8943540674 | 0.0 | 0.0
|$10^{6}$| 4358898.943540674  | 0.2249281406402588 | 4358898.943540674 | 0.0 | 0.0
|$10^{7}$| 43588989.43540674  | 2.2114503383636475 | 43588989.43540674 | 0.0 | 0.0
|$10^{8}$| 435889894.3540674  | 22.463025093078613 | 435889894.3540674 | 0.0 | 0.0


**Analysis**

In comparing the naive algorithm to the compensated algorithm it can be seen that the naive algorithm is faster in terms of run time. When running the program on my computer the naive algorithm took around 14 seconds to finish to the tenth power. However, the compensated algorithm finished in about 22 seconds with 0.0 for both the absoulte error and relative error. The naive algorithm is faster with the expense of having more errors. In our class presentation we discussed that, "actual errors are significantly less than the theoretical bounds on those errors". According to our data for compensated algorithm the absolute and relative error is less than the naive algorithm, thus telling us that the errors are consistent. The naive algorithm errors is increasing  when N increases whereas the compensated algorithm errors is static when N increases.